app/custom_auth.py

```diff
@@ -0,0 +1,89 @@
+    HTTP_X_AUTH_PROXY_USER : User 
+    HTTP_X_AUTH_PROXY_GROUPS: Admin,Editor
+
+    Summary
+    -------
+    (1) If HTTP header HTTP_X_AUTH_PROXY_USER with the user name is not passed no further processing is done by the middleware.
+    (2) If HTTP header with the user name is passed then:
+        (2.1) Authenciate the user if already exist in the database. 
+        (2.2) Create user if does not exist in the database and then autenicate the user. 
+    (3) If IS_SUPER flag is set to True in the settings.py then ingnore HTTP_X_AUTH_PROXY_GROUPS and grant the user Admin level privileges
+    (4) If HTTP_X_AUTH_PROXY_GROUPS header is present and contains Admin group then grant the user Admin level privileges. 
+    (5) If HTTP_X_AUTH_PROXY_GROUPS header is absent or does not contain the Admin group then revoke user's Admin level privleges.  
+    (6) User is now signed in and session cookies are set.  
+    ...
+    """
+    def __init__(self, get_response):
+        self.get_response = get_response
+     
+    def __call__(self, request):  
+        if not request.user.is_authenticated:
+            try:
+                username = request.META[settings.HEADER_AUTH_PROXY_USER]
+            except KeyError:
+                logger.exception(f"User name header {settings.HEADER_AUTH_PROXY_USER} not found. Continue without user authenication." )
+                return self.get_response(request)
+
+            # authenticate() method automatically creates the user if it is not already exists in the system.
+            user = auth.authenticate(remote_user=username)
+            process_user_groups(user,request.META)
+                
+            if user:
+                request.user = user
+                auth.login(request, user)
+
+        return self.get_response(request)
+
+def process_user_groups(user,headers):
+        groups_from_header = get_user_groups_from_header(headers)
+        admin_group_name = settings.ADMIN_GROUP_NAME
+        admin_group = get_admin_group(admin_group_name)
+        
+        # Case 1: IS_SUPER flag takes precedence over HTTP_X_AUTH_PROXY_GROUPS header.
+        if settings.IS_SUPER:
+                if not user.groups.filter(name=admin_group_name).exists():
+                    user.groups.add(admin_group)
+                user.is_superuser = settings.IS_SUPER
+                user.save()
+                return 
+        
+        # Case 2: HTTP_X_AUTH_PROXY_GROUPS header contains Admin group. 
+        if  groups_from_header and admin_group_name in groups_from_header:
+                if not user.groups.filter(name=admin_group_name).exists():
+                    user.groups.add(admin_group)
+                user.is_superuser = True
+                user.save()
+
+        # Case 3: HTTP_X_AUTH_PROXY_GROUPS header is absent or does not contain Admin group.
+        if not groups_from_header or not admin_group_name in groups_from_header:
+                if user.groups.filter(name=admin_group_name).exists():
```

Comment:
As discussed, we're not implementing groups for now, so let's focus on just the superuser flag for now. In this way we can collapse both of these if statements into one:

```py
user.is_superuser = groups_from_header and admin_group_name in groups_from_header
user.save()
```